service/src/java/org/apache/hive/service/cli/thrift/ThriftBinaryCLIService.java

```diff
@@ -113,43 +113,66 @@ protected void initServer() {
           Metrics metrics = MetricsFactory.getInstance();
           if (metrics != null) {
-            try {
-              metrics.decrementCounter(MetricsConstant.OPEN_CONNECTIONS);
-            } catch (Exception e) {
-              LOG.warn("Error Reporting JDO operation to Metrics system", e);
-            }
+            metrics.decrementCounter(MetricsConstant.OPEN_CONNECTIONS);
           }
-          ThriftCLIServerContext context = (ThriftCLIServerContext) serverContext;
-          SessionHandle sessionHandle = context.getSessionHandle();
-          if (sessionHandle != null) {
-            LOG.info("Session disconnected without closing properly. ");
+
+          final ThriftCLIServerContext context = (ThriftCLIServerContext) serverContext;
+          final Optional<SessionHandle> sessionHandle = context.getSessionHandle();
+
+          if (sessionHandle.isPresent()) {
+            // Normally, the client should politely inform the server it is
+            // closing its session with Hive before closing its network
+            // connection. However, if the client connection dies for any reason
+            // (load-balancer round-robin configuration, firewall kills
+            // long-running sessions, bad client, failed client, timed-out
+            // client, etc.) then the server will close the connection without
+            // having properly cleaned up the Hive session (resources,
+            // configuration, logging etc.). That needs to be cleaned up now.
+            LOG.warn(
+                "Client connection bound to {} unexpectedly closed: closing this Hive session to release its resources. "
+                    + "The connection processed {} total messages during its lifetime of {}ms. Inspect the client connection "
+                    + "for time-out, firewall killing the connection, invalid load balancer configuration, etc.",
+                sessionHandle, context.getMessagesProcessedCount(), context.getDuration().toMillis());
             try {
-              boolean close = cliService.getSessionManager().getSession(sessionHandle).getHiveConf()
+              final boolean close = cliService.getSessionManager().getSession(sessionHandle.get()).getHiveConf()
                   .getBoolVar(ConfVars.HIVE_SERVER2_CLOSE_SESSION_ON_DISCONNECT);
-              LOG.info((close ? "" : "Not ") + "Closing the session: " + sessionHandle);
               if (close) {
-                cliService.closeSession(sessionHandle);
+                cliService.closeSession(sessionHandle.get());
+              } else {
+                LOG.warn("Session not actually closed because configuration {} is set to false",
+                    ConfVars.HIVE_SERVER2_CLOSE_SESSION_ON_DISCONNECT.varname);
               }
             } catch (HiveSQLException e) {
-              LOG.warn("Failed to close session: " + e, e);
+              LOG.warn("Failed to close session", e);
+            }
+          } else {
```

Comment:
Thanks for the work, @belugabehr .  l left some minor ideas,  others look good to me...
Nit:  maybe we should remove ```There is no session handle because the client gracefully closed the session```

Reply:
I was thinking about this as well, but there is no logging when the context is created, so it's a bit confusing to log when the context is destroyed.  Perhaps this can be more holistically considered in another JIRA.  For now, it's best to log nothing if everything is OK.